collector/rapl_linux.go

```diff
@@ -0,0 +1,89 @@
+// Copyright 2019 The Prometheus Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// +build !norapl
+
+package collector
+
+import (
+	"strconv"
+
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/procfs/sysfs"
+)
+
+type raplMetric struct {
+	descriptor *prometheus.Desc // prometheus descriptor
+	raplZone   sysfs.RaplZone   // RAPL zone for the metric
+}
+
+type raplCollector struct {
+	metrics []raplMetric
+}
+
+func init() {
+	registerCollector("rapl", defaultEnabled, NewRaplCollector)
+}
+
+// NewRaplCollector returns a new Collector exposing RAPL metrics.
+func NewRaplCollector() (Collector, error) {
+	fs, err := sysfs.NewFS(*sysPath)
+	if err == nil {
+		collector := raplCollector{}
+		// Error intentionally unchecked, nil zones are fine when platform
+		// doesn't have powercap files present. Update will return immediately
+		// for nil zones.
+		zones, _ := sysfs.GetRaplZones(fs)
```

Comment:
Sorry, I should have noticed this earlier in the review.

This is calling the metric update within the init function. This means it will only get called once at node_exporter startup.

What you want to do is put the `fs` handle into the `raplCollector` struct, and call `GetRaplZones` from the `Update()` function. Take a look at `thermal_zone_linux.go` or `cpufreq_linux.go` for examples.

Reply:
The GetRaplZones is intentionally only called once, as it searches through the filesystem and tries to find the rapl zone names. The directory structure is not expected to change after init.

During Update the stored zones are then read via raplZone.GetEnergyMicrojoules. This is an optimization which makes the Update faster and simpler. There is no need anymore at that point to parse through the filesystem, it is enough to just read the values during Update.

Reply:
Even if it's not likely to change, the kernel is modular, and could change out from under the exporter. We intentionally want to scan for the list of zones at each scrape. If a user were to unload `intel_rapl`, the exporter would continue to try scanning for non-existent data.

Reply:
I do find that a rather unlikely use-case, but if you really want it to support dynamic module loading and unloading, then it does need a rewrite, and that directory structure needs to be parsed during each and every scrape.

Reply:
Yes, we need to support it.

Reply:
Ok, done and refactored. I manually verified the functionality while scraping by running "modprobe -r intel_rapl" -> rapl related metrics vanish, and node_exporter keeps functioning fine. After "modprobe intel_rapl" the metrics reappear.